examples/experimental/trial/gaea_nas/model_def.py

```diff
@@ -0,0 +1,142 @@
+                criterion,
+                self.hparams.nodes,
+            )
+        )
+
+        # Initialize the optimizers and learning rate scheduler.
+        self.ws_opt = self.context._Optimizer(
+            torch.optim.SGD(
+                self.model.ws_parameters(),
+                self.hparams.learning_rate,
+                momentum=self.hparams.momentum,
+                weight_decay=self.hparams.weight_decay,
+            )
+        )
+        self.arch_opt = self.context._Optimizer(
+            EG(
+                self.model.arch_parameters(), 
+                self.hparams.arch_learning_rate,
+                lambda p: p / p.sum(dim=-1, keepdim=True)
+            )
+        )
+
+        self.lr_scheduler = self.context._LRScheduler(
+            lr_scheduler=CosineAnnealingLR(
+                self.ws_opt, 
+                self.hparams.scheduler_epochs,
+                self.hparams.min_learning_rate
+            ),
+            step_mode=LRScheduler.StepMode.STEP_EVERY_EPOCH,
+        )
+
+    def build_training_data_loader(self) -> DataLoader:
+        """
+        For bi-level NAS, we'll need each instance from the dataloader to have one image 
+        for training shared-weights and another for updating architecture parameters.
+        """
+        train_transform, _ = data_transforms_cifar10()
+        train_data = dset.CIFAR10(
+            root=self.data_dir, train=True, download=True, transform=train_transform
+        )
+        bilevel_data = BilevelDataset(train_data)
+
+        self.train_data = bilevel_data
+
+        train_queue = DataLoader(
+            bilevel_data,
+            batch_size=self.context.get_per_slot_batch_size(),
+            shuffle=True,
+            num_workers=2
+        )
+        return train_queue
+
+    def build_validation_data_loader(self) -> DataLoader:
+        _, valid_transform = data_transforms_cifar10()
+        valid_data = dset.CIFAR10(
+            root=self.data_dir, train=False, download=True, transform=valid_transform
+        )
+        valid_queue = DataLoader(
+            valid_data,
+            batch_size=self.context.get_per_slot_batch_size(),
+            shuffle=False,
+            num_workers=2
+        )
+        return valid_queue
+
+    def train_batch(
+        self, batch: TorchData, model: nn.Module, epoch_idx: int, batch_idx: int
+    ) -> Dict[str, torch.Tensor]:
+        if epoch_idx != self.last_epoch:
+            self.train_data.shuffle_val_inds()
+        self.last_epoch = epoch_idx
+        x_train, y_train, x_val, y_val = batch
+
+        # Train shared-weights
+        for a in self.model.arch_parameters():
+            a.requires_grad=False
+        for w in self.model.ws_parameters():
+            w.requires_grad=True
+        loss = self.model._loss(x_train, y_train)
+        self.context._backward(loss)
+        self.context._step_optimizer(self.ws_opt)
+
+        # Train arch parameters
+        for a in self.model.arch_parameters():
```

Comment:
ws_opt and arch_opt optimize disjoint sets of parameters of the model.  the optimizer.zero_grad() call at the end of _step_optimizer(self.ws_opt) does not zero out the gradients computed for the arch_parameters.  

without specifying requires_grad for the two sets of parameters, this throws an error for calling backward multiple times.  it was surprising that self.model.zero_grad() did not solve the issue.  i think allowing users to call zero_grad on the model with solve this issue and make it easier to optimize disjoint sets of variables of the same model.

Reply:
@shiyuann

Reply:
> ws_opt and arch_opt optimize disjoint sets of parameters of the model. the optimizer.zero_grad() call at the end of _step_optimizer(self.ws_opt) does not zero out the gradients computed for the arch_parameters.

I noticed that you set the aggregation_frequency to be 1. With this setting, `step_optimizer` should call `optimizer.zero_grad()` for users and all the parameters associated with that optimizer should clear out their gradients. Wonder if `optimizer.zero_grad()` was actually called.

> Without specifying requires_grad for the two sets of parameters, this throws an error for calling backward multiple times. it was surprising that self.model.zero_grad() did not solve the issue.

I noticed that you were only using parallel training and not AMP.

This is a constraint with horovod because it needs us to tell how many backward passes are done for each optimizer before running `optimizer.step()` on it. It has this constraint because they want the users to write the right code to avoid unnecessary across-GPU communication in distributed training. There are three methods to design the interface to satisfy this constraint.

1. Letting users tell us how many times they want to call `backward` on losses before they call `step_optimizer` each time in advance. This will make the interface ugly and hard to understand.

2. Assume users call `step_optimizer` right after `backward(loss)` each time and assume users will set `requires_grad` for those parameters they want to calculate gradients on. This loses the flexibility to do manual gradient accumulation but helps users write the right code to avoid unnecessary across-GPU communication.

3. Work around horovod to know how many times `backward` is called on losses for each parameter. This makes calling `backward` and `step_optimizer` flexible enough but users might find d-train not as effective as they thought just because they write the inefficient code.

The 2nd method is what we chose intentionally. So rather than exposing `zero_grad` we probably should explain to users more clearly how we want them to write their code.

Reply:
I am still looking into if we could support local gradient accumulation and avoid doing unnecessary communication. Supporting `zero_grad` generally looks reasonable to me.

Reply:
Added a ticket https://determinedai.atlassian.net/browse/DET-3513 and add it to #ml-ag meeting.